packages/dropcaps.lua

```diff
@@ -0,0 +1,92 @@
+SILE.require("packages/rebox")
+SILE.require("packages/raiselower")
+
+local shapeHbox = function (options, content)
+  -- Clear irrelevant values before passing to font
+  options.lines, options.join, options.raise, options.shift, options.color = nil, nil, nil, nil, nil
+  SILE.call("noindent")
+  local hbox = SILE.call("hbox", {}, function ()
+    SILE.call("font", options, content)
+  end)
+  table.remove(SILE.typesetter.state.nodes)
+  return hbox
+end
+
+-- This implementation relies on the hangafter and hangindent features of Knuth's line-breaking algorithm.
+-- These features in core line breaking algorithm supply the blank space in the paragraph shape but don't fill it with anything.
+SILE.registerCommand("dropcap", function (options, content)
+  local lines = SU.cast("integer", options.lines or 3)
+  local join = SU.boolean(options.join, false)
+  local standoff = SU.cast("measurement", options.standoff or "1spc")
+  local raise = SU.cast("measurement", options.raise or 0)
+  local shift = SU.cast("measurement", options.shift or 0)
+  local size = SU.cast("measurement or nil", options.size or nil)
+  local color = options.color
+  options.size = nil -- we need to measure the "would have been" size before using this
+
+  if color then SILE.require("packages/color") end
+
+  -- We want the drop cap to span over N lines, that is N - 1 baselineskip + the height of the first line.
+  -- Note this only works for the default linespace mechanism.
+  -- We determine the height of the first line by measuring the size the initial content *would have* been.
+  -- This gives the font some control over its relative size, sometimes desired sometimes undesired.
+  local tmpHbox = shapeHbox(options, content)
+  local extraHeight = SILE.measurement((lines - 1).."bs"):tonumber()
+  local targetHeight = tmpHbox.height:tonumber() + extraHeight
```

Comment:
An option here could be interesting... Let me try to explain what I mean, the target height currently is:

![image](https://user-images.githubusercontent.com/18075640/135098484-02aaf5ec-66f2-4212-b746-6f9b8f49ab0c.png)

But as can been seen in this example, the height the initial would have taken is bigger than the line content (in the main text font, cf. especially the "V" later on). That can be checked _without_ dropped initial.

![image](https://user-images.githubusercontent.com/18075640/135099268-611d1b41-aa81-492c-8322-ded5b740bed2.png)

Of course, the first line content can be anything, use several fonts that don't match in height, even contain other things than raw text... So it's unlikely there is a general solution here. On the other hand, the main usage should be for text, and wouldn't it be nice if we could get:

![image](https://user-images.githubusercontent.com/18075640/135101125-a23349e7-0338-4585-a3c2-4241315c2323.png)

Here above, the initial fits well with the line top (aligned on the V). What I did here is obviously not general:
```
  local targetHeight = SILE.shaper:measureChar("V").height + extraHeight
```
If you remember my initial commit, that's something I had tried (but measuring a "i" rather than a "V" - kind of _ad hoc_ I admit, and even less general)

Still, I think we could try to address it. The default would still be `tmpHbox.height:tonumber()`, but say for instance we have an option `aligntopon=V` (I can't think of a better name)? Or a better way to find out the height of the subsequent characters.

Perhaps even more general, a command to store some content such as:

```
\dropcap:alignement{any content I want}
``` 
And then
```
  local lineAdjustBox = dpContentForVerticalAlign and shapeHbox({}, dpContentForVerticalAlign) or tmpHbox
  local targetHeight = lineAdjustBox.height + extraHeight
```
So one would have more flexibility adjust the vertical top alignment. E.g.

![image](https://user-images.githubusercontent.com/18075640/135105805-e618e9fe-e602-472f-b498-a7f992a55ca9.png)

Here I decided to have the initial top-aligned to the smallcaps (which in _this_ font, are of course smaller than a capital, but slightly higher then an "x").

I am not sure how to formalize it (good options or commands for it), and maybe there would be even a nicer way to achieve that result. But it sure would look good to have this kind if flexibility, no? 

EDIT: Fixed a few typos in my comment.

Reply:
Yes I understand you viewpoint..After a second thought, why not just a "first line factor" (default 1.0) applied to that `tmpHbox.height:tonumber()` i(and the responsibility of computing the ratio is left to user) ?

On the other point, I suppose you mean the (bottom, now) alignment is not perfect with the last indented line, do you? And that the baselineskip is only indicative as the TeX paragraph building can stretch/expand the actual lineskip, right? I'm bothered by it too (not only for dropcaps). Started to look into alternate line spacing from the "linespacing" package and into ##1118, trying to understand how it works there...

Reply:
The optional `size` parameter can already be used if the user wants to calculate their own size, but I see what you mean about a relative scale factor being more ergonomic for some use cases. I just added `scale` (defaulting to 1.0). The [tweaked example shows usage](https://github.com/sile-typesetter/sile-typesetter.github.io/pull/26/commits/0847d4b0d8522b025113ffb4cf2929b90b5e9559), here is the relevant output of that use of scale:

![image](https://user-images.githubusercontent.com/173595/135241261-50a9b2d2-f694-4cb3-bd9c-8261a1df2450.png)

Reply:
As for the bottom alignment, I think I now get it: We use "bs" units converted an to absolute number, but
- "bs" is derived from "document.baselineskip"
- In the TeX linespacing, "document.baselineskip" as some stretch, and "document.lineskip" may also occur - resulting in an imperfect alignment
- The other linespacing mechanisms don't at all rely on "document.baselineskip", so using "bs" is broken by nature in these cases.

This raises two main questions:
- What is supposed to be a "bs" unit when the linespacing is not the default (TeX) one?
- Either the dropcap must be computed _after_ the paragraph is built (as you implied above), or it seems acceptable by definition that such a paragraph should have stretch between baselines.

Reply:
For my current project (which uses linespace method `fixed`), I set the baselineskip value to be a copy of what I used for fixed:

```lua

-- Set so bs unit works. Also note dropcaps would be based on this if we didn't manually set them.
SILE.settings.set("document.baselineskip", SILE.settings.get("linespacing.fixed.baselinedistance"))
```

Obviously that's a bit less than ideal, but having this working for the (current) default leading calculator seems like a good enough place to start. We can add detection and support for others later.

Honestly I think the solution to this is to refactor the entire linespacing system so that all of the methods are modular drop in packages that load their own settings and can redefine some generic unit for use in situations like this and/or provide a callback to get absolute values from actual lines rather than predictions. That can go in #1118.